jormungandr/src/network/p2p/topology.rs

```diff
@@ -2,171 +2,207 @@
+    #[serde(serialize_with = "serialize_display")]
+    pub id: NodeId,
+    pub address: Address,
+    pub last_update: SystemTime,
+    pub quarantined: Option<SystemTime>,
+    pub subscriptions: Vec<(String, String)>,
+}
 
-    fn set_policy(mut self, policy: PolicyConfig) -> Self {
-        self.topology.set_policy(Policy::new(
-            policy,
-            span!(parent: &self.span, Level::TRACE, "sub_task", kind = "policy"),
-        ));
-        self
+impl PartialEq for ProfileInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id && self.address == other.address
     }
+}
 
-    /// set all the default poldercast modules (Rings, Vicinity and Cyclon)
-    fn set_poldercast_modules(mut self) -> Self {
-        self.topology.add_layer(Rings::default());
-        self.topology.add_layer(Vicinity::default());
-        self.topology.add_layer(Cyclon::default());
-        self
+impl Hash for ProfileInfo {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.id.hash(state);
+        self.address.hash(state);
     }
+}
 
-    fn set_custom_modules(mut self, config: &Configuration, rng: ChaChaRng) -> Self {
-        if let Some(size) = config.max_unreachable_nodes_to_connect_per_event {
-            self.topology
-                .add_layer(custom_layers::RandomDirectConnections::with_max_view_length(size));
-        } else {
-            self.topology
-                .add_layer(custom_layers::RandomDirectConnections::default());
+impl From<&Arc<Profile>> for ProfileInfo {
+    fn from(other: &Arc<Profile>) -> Self {
+        Self {
+            id: other.id(),
+            address: other.address(),
+            last_update: other.last_update().to_system_time().into(),
+            quarantined: None,
+            subscriptions: other
+                .subscriptions()
+                .iter()
+                .map(|s| (s.topic().to_string(), format!("{:?}", s.interest_level())))
+                .collect(),
         }
+    }
+}
 
-        self.topology.add_layer(PreferredListLayer::new(
-            config.layers.preferred_list.clone(),
-            rng,
-        ));
+pub struct View {
+    pub peers: Vec<Arc<Profile>>,
+}
 
-        self
-    }
+struct Inner {
+    topology: Topology,
+    quarantine: Quarantine,
+}
 
-    fn build(self) -> P2pTopology {
-        P2pTopology {
-            lock: RwLock::new(self.topology),
-        }
-    }
+/// object holding the P2pTopology of the Node
+pub struct P2pTopology {
+    lock: RwLock<Inner>,
 }
 
 impl P2pTopology {
-    pub fn new(config: &Configuration, span: Span, rng: ChaChaRng) -> Self {
-        Builder::new(config.profile.clone(), span)
-            .set_poldercast_modules()
-            .set_custom_modules(&config, rng)
-            .set_policy(config.policy.clone())
-            .build()
+    pub fn new(config: &Configuration) -> Self {
+        // FIXME: How should we handle cases where the is not listen set? Can a node just receive?
```

Comment:
@mzabaluev we need your clarification on that.

Reply:
I guess this is where the public address thing comes to the fore.
My first approach would be to:
- Always require to listen on some socket, just to simplify all code paths;
- If the public address is not set, try to create a dummy profile with the local binding address retrieved from `TcpListener::local_addr` and never advertise it in gossip?

Reply:
What about we use some constant that can't be for sure used in a network (even a private one) like `127.0.0.1:0` if the public address is not defined and then manually filter outgoing gossips? On a side note, what would happen if a node tries to connect to itself (e.g because a malicious user create another routing for the public address of one of the nodes)?

As for the requirement on listen address, I think the only place where it is used is when starting the grpc server, so we can simply not do that if it's not set (and we already do that), I don't see how always enforcing one would make things simpler.